src/Microsoft.AspNetCore.Mvc.Core/Routing/EndpointRoutingUrlHelper.cs

```diff
@@ -51,46 +52,41 @@ public override string Action(UrlActionContext urlActionContext)
                 throw new ArgumentNullException(nameof(urlActionContext));
             }
 
-            var valuesDictionary = GetValuesDictionary(urlActionContext.Values);
+            var values = GetValuesDictionary(urlActionContext.Values);
 
             if (urlActionContext.Action == null)
             {
-                if (!valuesDictionary.ContainsKey("action") &&
+                if (!values.ContainsKey("action") &&
                     AmbientValues.TryGetValue("action", out var action))
                 {
-                    valuesDictionary["action"] = action;
+                    values["action"] = action;
                 }
             }
             else
             {
-                valuesDictionary["action"] = urlActionContext.Action;
+                values["action"] = urlActionContext.Action;
             }
 
             if (urlActionContext.Controller == null)
             {
-                if (!valuesDictionary.ContainsKey("controller") &&
+                if (!values.ContainsKey("controller") &&
                     AmbientValues.TryGetValue("controller", out var controller))
                 {
-                    valuesDictionary["controller"] = controller;
+                    values["controller"] = controller;
                 }
             }
             else
             {
-                valuesDictionary["controller"] = urlActionContext.Controller;
+                values["controller"] = urlActionContext.Controller;
             }
 
-            var successfullyGeneratedLink = _linkGenerator.TryGetLink(
-                ActionContext.HttpContext,
-                valuesDictionary,
-                out var link);
-            if (!successfullyGeneratedLink)
-            {
-                //TODO: log here
 
-                return null;
-            }
-
-            return GenerateUrl(urlActionContext.Protocol, urlActionContext.Host, link, urlActionContext.Fragment);
+            var path = _linkGenerator.GetPathByRouteValues(
+                ActionContext.HttpContext,
+                routeName: null,
+                values,
+                new FragmentString(urlActionContext.Fragment == null ? null : "#" + urlActionContext.Fragment));
+            return GenerateUrl(urlActionContext.Protocol, urlActionContext.Host, path);
```

Comment:
There's an impedance mismatch between what link generator and routing do that results in this code changing without much good opportunity for code reuse.

1. LinkGenerator always returns a Path + Query + Fragment, not a virtual path. 
2. MVC exposes APIs that allow you to generate a URL with a different host/scheme than what's on the request.
3. LinkGenerator doesn't allow you to **both** accept a host/scheme and use ambient values.

I think we should consider addressing the 3rd point since it's just a general usability problem.

Reply:
But not in this PR :)